packages/app-check/lib/index.js

```diff
@@ -40,6 +40,16 @@ const namespace = 'appCheck';
 const nativeModuleName = 'RNFBAppCheckModule';
 
 class FirebaseAppCheckModule extends FirebaseModule {
+  constructor(...args) {
+    super(...args);
+
+    this.emitter.addListener(this.eventNameForApp('appCheck_token_changed'), event => {
+      this.emitter.emit(this.eventNameForApp('onAppCheckTokenChanged'), event);
+    });
+
+    this.native.addAppCheckListener();
```

Comment:
Interesting - this adds a listener even if nothing is listening, but similar to remote-config's `onConfigUpdated` functionality, the upstream firebase-js-sdk docs indicate that an app may register zero or more listeners for a single app and they will all be called for the event

We can do that at the javascript level by allowing multiple listeners per app and just keeping track of how many listeners we have

So this should maintain a count of listeners at the javascript level, only subscribe to the native level if adding the first javascript level listener, and only remove the native listener from the native level if removing the last listener, similar to remote-config

https://github.com/invertase/react-native-firebase/blob/ea38a20988fbbb030d1a28d1ae83f627b47d615d/packages/remote-config/lib/index.js#L312-L331

Reply:
Hmm - well one of the wonderful things about mirroring firebase-js-sdk behavior is that we don't have to make judgement calls about whether something is useful or not. Our criteria is fidelity - we need to behave the *same* as firebase-js-sdk, regardless of opinion on the functionality.

So our job is really just to make sure we understand the functionality, then we implement it. Here is what I read in the related docs:

> Registers a listener to changes in the token state. There can be more than one listener registered at the same time for one or more App Check instances. The listeners call back on the UI thread whenever the current token associated with this App Check instance changes.

ref: https://firebase.google.com/docs/reference/js/app-check.md#ontokenchanged

I read that - if I reduce it slightly but I believe in keeping with the logical conditions: "there can be more than one listener registered at the same time for one App Check instances". I believe that is still logically consistent because one branch of an or clause may be eliminated without altering the meaning, if the remaining clauses return the same value ("one instance" === true).

So I believe we need to support multiple listeners for one app. Perhaps someone has a use case they identified while designing the API? It would be similar to the remote-config listener API they implemented, I believe they intend to support multiple listeners for the same app.

Then concretely, they specifically test that multiple listeners for the same app work:

https://github.com/firebase/firebase-js-sdk/blob/2d0a9f5fd921568bc76fcdee325b9ab5e6be8a58/packages/app-check/src/api.test.ts#L336-L337

So in order to have fidelity with the firebase-js-sdk API and continue our intended status as a drop-in replacement, we need to handle it as well

The remote-config onConfigUpdated implementation has the useful bits of code to implement it I believe and should be pretty easy to do here ?